tests/functionaltests/Tests/UIKitTests/UIViewControllerTests.mm

```diff
@@ -0,0 +1,310 @@
+            EXPECT_TRUE(sel_isEqual([presenting lastEvent], @selector(viewDidAppear:)));
+        }
+    }
+}
+
+- (void)viewWillDisappear:(BOOL)animated {
+    [super viewWillDisappear:animated];
+
+    EXPECT_TRUE(sel_isEqual([self lastEvent], @selector(viewDidAppear:)));
+    _lastEvent = _cmd;
+
+    if ([self presentedViewController]) {
+        // About to be covered by presented modal
+
+        EXPECT_FALSE([self isBeingDismissed]);
+
+        ModalEventsTestViewController* presented = (ModalEventsTestViewController*)[self presentedViewController];
+
+        EXPECT_TRUE(_hidesParent(presented.modalPresentationStyle));
+
+        // We should receive viewWillDisappear: before presented receives viewWillAppear:
+        // Ensure presented is displaying for the first time (or was previously dismissed)
+        EXPECT_TRUE(sel_isEqual([presented lastEvent], @selector(viewDidLoad)) ||
+                    sel_isEqual([presented lastEvent], @selector(viewDidDisappear:)));
+    } else {
+        // About to be dismissed
+
+        EXPECT_TRUE([self isBeingDismissed]);
+
+        EXPECT_NE([self presentingViewController], nil);
+
+        if ([[self presentingViewController] isKindOfClass:[ModalEventsTestViewController class]]) {
+            ModalEventsTestViewController* presenting = (ModalEventsTestViewController*)[self presentingViewController];
+
+            if (_hidesParent(self.modalPresentationStyle)) {
+                // We should receive viewWillDisappear: before presenting receives viewWillAppear:
+                EXPECT_TRUE(sel_isEqual([presenting lastEvent], @selector(viewDidDisappear:)));
+            } else {
+                EXPECT_TRUE(sel_isEqual([presenting lastEvent], @selector(viewDidAppear:)));
+            }
+        }
+    }
+}
+
+- (void)viewDidDisappear:(BOOL)animated {
+    [super viewDidDisappear:animated];
+
+    EXPECT_TRUE(sel_isEqual(_lastEvent, @selector(viewWillDisappear:)));
+    _lastEvent = _cmd;
+
+    if ([self presentedViewController]) {
+        // Modal is covering us.
+
+        EXPECT_FALSE([self isBeingDismissed]);
+
+        EXPECT_NE([self presentingViewController], nil);
+
+        ModalEventsTestViewController* presented = (ModalEventsTestViewController*)[self presentedViewController];
+
+        EXPECT_TRUE(_hidesParent(presented.modalPresentationStyle));
+
+        // presented should receive viewDidAppear: first.
+        EXPECT_TRUE(sel_isEqual([presented lastEvent], @selector(viewDidAppear:)));
+    } else {
+        // We're a dismissed modal.
+        EXPECT_EQ([self presentingViewController], nil);
+
+        EXPECT_TRUE([self isBeingDismissed]);
+    }
+}
+
+@end
+
+class UIViewControllerModalTest : public ::testing::TestWithParam<::testing::tuple<UIModalPresentationStyle, WOCOperationMode, BOOL>> {
+protected:
+    virtual void SetUp() {
+        FunctionalTestSetupUIApplication();
+    }
+
+    virtual void TearDown() {
+        FunctionalTestCleanupUIApplication();
+    }
```

Comment:
sounds good; thanks!

---
In reply to: [110289530](https://github.com/Microsoft/WinObjC/pull/2194#discussion_r110289530) [](ancestors = 110289530)